broker/alert_ingestion/valid_schemas/gen_valid_schema.py

```diff
@@ -0,0 +1,195 @@
+-------------
+
+    Survey versions that have already been configured are registered in the
+    ``format_funcs`` dict in the ``write_valid_schema()`` function. To
+    configure a new survey version, define a new function that accepts the
+    schema from the survey's Avro file (the schema to be corrected) as a dict
+    and returns the corrected schema as a dict. See ``_fix_schema_ZTF_v3_3()``
+    for an example. Register the new function in the ``format_funcs`` dict.
+    Then generate the pickle file as follows.
+
+.. code-block:: python
+   :linenos:
+
+   import gen_valid_schema as gvs
+
+   fin = <path to Avro file>
+   survey, version = 'ztf', 3.3
+   valid_schema = gvs.write_valid_schema(fin, survey, version)
+
+"""
+
+from pathlib import Path
+import logging
+import pickle
+import json
+import fastavro
+
+log = logging.getLogger(__name__)
+
+
+def write_valid_schema(fin: str, survey: str, version: float) -> dict:
+    """ Corrects an Avro file schema to comply with the strict validation
+    requirements of BigQuery. Writes the corrected schema to a pickle file and
+    also returns it as a dict.
+
+    Args:
+        fin         : Path to alert Avro file.
+        survey      : The name of the survey the alert is from.
+        version     : The schema version of the alert is from.
+
+    Returns:
+        schema  : corrected schema
+    """
+
+    format_funcs = {
+        'ztf': {
+            3.3: _fix_schema_ZTF_v3_3
+        }
+    }
+
+    fout = Path(__file__).resolve().parent / f'{survey}_v{version}.pkl'
+
+    schema, __ = _load_Avro(fin) # load the file
+
+    try:
+        format_func = format_funcs.get(survey, {})[version]
+
+    except IndexError:
+        err_msg = f'Formatting not available for {survey} {version}'
+        log.error(err_msg)
+        raise RuntimeError(err_msg)
+
+    else:
+        valid_schema = format_func(schema) # get the corrected schema
+        _write_schema_as_dict(valid_schema, fout)
+
+    return valid_schema
+
+def _fix_schema_ZTF_v3_3(schema: dict):
+    """ Corrects the ZTF version 3.3 schema to comply with the strict Avro
+    validation requirements of BigQuery.
+
+    Args:
+        schema : Avro schema header, as returned by _load_Avro()
+    Returns:
+        schema : updated schema
+    """
+    for l1, l1_field in enumerate(schema['fields']): # l1_field is a dict
+
+        # do the top level fields
+        schema['fields'][l1] = _reverse_types(l1_field)
+
+        # do the nested candidate fields
+        if l1_field['name'] == 'candidate':
+            for l2, l2_field in enumerate(l1_field['type']['fields']):
+                schema['fields'][l1]['type']['fields'][l2] = _reverse_types(l2_field)
+
+        # do the nested prv_candidate fields
+        if l1_field['name'] == 'prv_candidates':
+            for l2, l2_field in enumerate(l1_field['type'][1]['items']['fields']):
+                schema['fields'][l1]['type'][1]['items']['fields'][l2] = _reverse_types(l2_field)
```

Comment:
I agree that it's difficult to follow... I didn't even like it when I wrote it. I had a hard time getting it to work when I defined intermediate variables, and I'm not sure why. Since the schema file is already generated we will never need to use this function again, so I'd prefer not to spend the time to change this and make sure it works.